autogpt_platform/backend/backend/data/graph.py

```diff
@@ -1112,55 +1095,104 @@ async def delete_graph(graph_id: str, user_id: str) -> int:
 
 
 async def validate_graph_execution_permissions(
-    graph_id: str, user_id: str, graph_version: Optional[int] = None
+    user_id: str, graph_id: str, graph_version: int, is_sub_graph: bool = False
 ) -> None:
     """
     Validate that a user has permission to execute a specific graph.
 
     This function performs comprehensive authorization checks and raises specific
     exceptions for different types of failures to enable appropriate error handling.
 
+    ## Logic
+    A user can execute a graph if any of these is true:
+    1. They own the graph and some version of it is still listed in their library
+    2. The graph is published in the marketplace and listed in their library
+    3. The graph is published in the marketplace and is being executed as a sub-agent
+
     Args:
         graph_id: The ID of the graph to check
         user_id: The ID of the user
-        graph_version: Optional specific version to check. If None (recommended),
-                      performs version-agnostic check allowing execution of any
-                      version as long as the graph is in the user's library.
-                      This is important for sub-graphs that may reference older
-                      versions no longer in the library.
+        graph_version: The version of the graph to check
+        is_sub_graph: Whether this is being executed as a sub-graph.
+            If `True`, the graph isn't required to be in the user's Library.
 
     Raises:
         GraphNotInLibraryError: If the graph is not in the user's library (deleted/archived)
         NotAuthorizedError: If the user lacks execution permissions for other reasons
     """
+    # Step 1: Check if user owns this graph
+    user_owns_graph = (
+        graph := await AgentGraph.prisma().find_unique(
+            where={"graphVersionId": {"id": graph_id, "version": graph_version}}
+        )
+    ) and graph.userId == user_id
 
-    # Step 1: Check library membership (raises specific GraphNotInLibraryError)
-    where_clause: LibraryAgentWhereInput = {
-        "userId": user_id,
-        "agentGraphId": graph_id,
-        "isDeleted": False,
-        "isArchived": False,
-    }
+    # Step 2: Check if agent is in the library *and not deleted*
+    user_has_in_library = (
+        await LibraryAgent.prisma().find_first(
+            where={
+                "userId": user_id,
+                "agentGraphId": graph_id,
+                "isDeleted": False,
+            }
+        )
+        is not None
+    )
```

Comment:
nit: this two changes are being queries eitherway, maybe we should just run the two queries at the same time and asyncio.gather it then evaluate the result? Or the other option is  pulling LibraryAgent along with the AgentGraph (included in the query) so we do it only once and validate the value ?